XMonad/Util/Loggers.hs

```diff
@@ -203,12 +203,47 @@ logTitlesOnScreen sid formatFoc formatUnfoc = (`withScreen` sid) $ \screen -> do
                  wins
                  winNames
 
+-- | Variant of 'logTitlesOnScreen' with support for urgent windows. To be used
+-- with 'XMonad.Hooks.UrgencyHook'.
+logTitlesOnScreen' :: ScreenId -> TitlesFormat' -> Logger
+logTitlesOnScreen' sid (TitlesFormat' formatFoc formatUnfoc formatUrg) = (`withScreen` sid) $ \screen -> do
+  let focWin = fmap W.focus . W.stack . W.workspace $ screen
+      wins = maybe [] W.integrate . W.stack . W.workspace $ screen
+  urgWins <- readUrgents
+  winNames <- traverse (fmap show . getName) wins
+  pure . Just
+    . unwords
+    $ zipWith
+      ( \w n ->
+          if Just w == focWin
+            then formatFoc n
+            else
+              if w `elem` urgWins
+                then formatUrg n
+                else formatUnfoc n
+      )
+      wins
+      winNames
+
 -- | Like 'logTitlesOnScreen', but directly use the "focused" screen
 -- (the one with the currently focused workspace).
-logTitles :: (String -> String) -> (String -> String) -> Logger
-logTitles formatFoc formatUnfoc = do
+logTitles :: TitlesFormat -> Logger
+logTitles (TitlesFormat formatFoc formatUnfoc) = do
   sid <- gets $ W.screen . W.current . windowset
-  logTitlesOnScreen sid formatFoc formatUnfoc
+  logTitlesOnScreen sid (TitlesFormat formatFoc formatUnfoc)
+
+-- | Variant of 'logTitles', but with support for urgent windows.
+logTitles' :: TitlesFormat' -> Logger
+logTitles' (TitlesFormat' formatFoc formatUnfoc formatUrg) = do
+  sid <- gets $ W.screen . W.current . windowset
+  logTitlesOnScreen' sid (TitlesFormat' formatFoc formatUnfoc formatUrg)
+
+type FocusedFormat = (String -> String)
+type UnfocusedFormat = (String -> String)
+type UrgentFormat = (String -> String)
+
+data TitlesFormat = TitlesFormat FocusedFormat UnfocusedFormat
+data TitlesFormat' = TitlesFormat' FocusedFormat UnfocusedFormat UrgentFormat
```

Comment:
This is probably easier if we use a record and give it a `Default` instance (users know how to deal with these)

```suggestion
data TitlesFormat' = TitlesFormat'
  { focusedFormat   :: String -> String
  , unfocusedFormat :: String -> String
  , urgentFormat    :: String -> String
  }

instance Default TitlesFormat' where
  def :: TitlesFormat'
  def = TitlesFormat' { … }
```

Reply:
Right, that makes much more sense. Thanks!